core/src/main/java/org/elasticsearch/common/Randomness.java

```
@@ -0,0 +1,119 @@
+            return create();
+        }
+    }
+
+    /**
+     * Provides a source of randomness that is reproducible when
+     * running under the Elasticsearch test suite, and otherwise
+     * produces a non-reproducible source of randomness. Reproducible
+     * sources of randomness are created when the system property
+     * "tests.seed" is set, is parseable as a hexadecimal long, and the
+     * security policy allows reading this system property. Otherwise,
+     * non-reproducible sources of randomness are created.
+     *
+     * @return a source of randomness
+     * @throws IllegalArgumentException if "tests.seed" is set but the
+     *                                  system property with that key
+     *                                  is not a string of hexadecimal
+     *                                  digits.
+     */
+    public static Random create() {
+        String property = null;
+        try {
+            property = (String)BootstrapInfo.getSystemProperties().get("tests.seed");
+        } catch (SecurityException ignored) {
+            // getting system properties requires wildcard read-write
+        }
+        if (property != null) {
+            long accumulator = 0;
+            for (char c : property.toCharArray()) {
+                accumulator = accumulator << 4;
+                if (c >= '0' && c <= '9')
+                    accumulator |= (c - '0');
+                else if (c >= 'a' && c <= 'f')
+                    accumulator |= (c - 'a' + 10);
+                else if (c >= 'A' && c <= 'F') {
+                    accumulator |= (c - 'A' + 10);
+                } else {
+                    throw new IllegalStateException("expected hexadecimal seed: " + property);
+                }
+            }
+            return createFromSeed(accumulator);
+        } else {
+            return createWithoutSeed();
+        }
+    }
+
+    private static Random createWithoutSeed() {
+        byte[] bytes = SR.generateSeed(8);
```

Can we throw an exception from here if we're in test code?